File: android/src/main/java/com/supermap/rnsupermap/JSMQTTClientSide.java
package com.supermap.rnsupermap;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.supermap.messagequeue.MQTTClient;

import java.util.Calendar;
import java.util.HashMap;
import java.util.Map;

/**
 * Created by Myself on 2017/5/16.
 */

public class JSMQTTClientSide extends ReactContextBaseJavaModule {
    public static final String REACT_CLASS = "JSMQTTClientSide";
    public static Map<String,MQTTClient> mClientList=new HashMap<String,MQTTClient>();
    MQTTClient mClient;

    public JSMQTTClientSide(ReactApplicationContext context){super(context);}

    public static String registerId(MQTTClient client){
        if(!mClientList.isEmpty()) {
            for(Map.Entry entry:mClientList.entrySet()){
                if(client.equals(entry.getValue())){
                    return (String)entry.getKey();
                }
            }
        }

        Calendar calendar=Calendar.getInstance();
        String id=Long.toString(calendar.getTimeInMillis());
        mClientList.put(id,client);
        return id;
    }

    @Override
    public String getName(){
        return REACT_CLASS;
    }

    @ReactMethod
    public void createObj(Promise promise){
        try{
            MQTTClient client = new MQTTClient();
            String _MQTTClientSideId = registerId(client);

            WritableMap map = Arguments.createMap();
            map.putString("_MQTTClientSideId",_MQTTClientSideId);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void create(String clientSideId,String URI,String userName,String password,String clientId,Promise promise){
        try{
            mClient = mClientList.get(clientSideId);
            boolean isConnection = mClient.create(URI,userName,password,clientId);

            WritableMap map = Arguments.createMap();
            map.putBoolean("isConnection",isConnection);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void sendMessage(String clientSideId,String topic,String message,Promise promise){
        try{
            mClient = mClientList.get(clientSideId);
            boolean isSent = mClient.sendMessage(topic,message);

            WritableMap map = Arguments.createMap();
            map.putBoolean("send",isSent);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void receiveMessage(String clientSideId,String topic,String message,Promise promise){
        try{
            mClient = mClientList.get(clientSideId);
            boolean isSent = mClient.sendMessage(topic,message);

            WritableMap map = Arguments.createMap();
            map.putBoolean("send",isSent);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void subscribe(String clientSideId,String topic,int qos,Promise promise){
        try{
            mClient = mClientList.get(clientSideId);
            boolean isSubscribe = mClient.subscribe(topic,qos);

            WritableMap map = Arguments.createMap();
            map.putBoolean("isSubscribe",isSubscribe);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void unsubscribe(String clientSideId,String topic,Promise promise){
        try{
            mClient = mClientList.get(clientSideId);
            boolean unSubscribe = mClient.unsubscribe(topic);

            WritableMap map = Arguments.createMap();
            map.putBoolean("unSubscribe",unSubscribe);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void suspend(String clientSideId,Promise promise){
        try{
            mClient = mClientList.get(clientSideId);
            mClient.suspend();

            promise.resolve("suspend");
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void resume(String clientSideId,Promise promise){
        try{
            mClient = mClientList.get(clientSideId);
            mClient.resume();

            promise.resolve("resume");
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void dispose(String clientSideId,Promise promise){
        try{
            mClient = mClientList.get(clientSideId);
            mClientList.remove(clientSideId);
            mClient.dispose();

            promise.resolve("dispose");
        }catch (Exception e){
            promise.reject(e);
        }
    }
}

File: ios/objNative/JSMQTTClientSide.m
//
//  JSMQTTClientSide.m
//  Supermap
//
//  Created by 王子豪 on 2017/5/16.
//  Copyright © 2017年 Facebook. All rights reserved.
//

#import "JSMQTTClientSide.h"

#import "SuperMap/MQTTClientSide.h"
#import "JSObjManager.h"
@implementation JSMQTTClientSide
RCT_EXPORT_MODULE();

//所有导出方法名
- (NSArray<NSString *> *)supportedEvents
{
    return @[@"com.supermap.RN.JSMQTTClientSide.receive_message1",@"com.supermap.RN.JSMQTTClientSide.receive_message2",@"com.supermap.RN.JSMQTTClientSide.receive_message3",@"com.supermap.RN.JSMQTTClientSide.receive_message4",@"com.supermap.RN.JSMQTTClientSide.receive_message5"];
}

 RCT_REMAP_METHOD(createObj,createObjWithresolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
     @try {
         MQTTClientSide* side = [[MQTTClientSide alloc]init];
         NSInteger nsKey = (NSInteger)side;
         [JSObjManager addObj:side];
         resolve(@{@"_MQTTClientSideId":@(nsKey).stringValue});
     } @catch (NSException *exception) {
         reject(@"MQTTClientSide",@"create Obj expection",nil);
     }
 }

RCT_REMAP_METHOD(create,createWithId:(NSString*)sideId URI:(NSString*)URI userName:(NSString*)name password:(NSString*)password clientId:(NSString*)clientId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        MQTTClientSide* side = [JSObjManager getObjWithKey:sideId];
        BOOL isConnection = [side create:URI username:name password:password clientId:clientId];
        NSNumber* num = [NSNumber numberWithBool:isConnection];
        resolve(@{@"isConnection":num});
    } @catch (NSException *exception) {
        reject(@"MQTTClientSide",@"create connection expection",nil);
    }
}

RCT_REMAP_METHOD(sendMessage,sendMessageWithId:(NSString*)sideId topic:(NSString*)topic message:(NSString*)message resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        MQTTClientSide* side = [JSObjManager getObjWithKey:sideId];
        BOOL sent = [side sendMessage:topic message:message];
        NSNumber* num = [NSNumber numberWithBool:sent];
        resolve(@{@"send":num});
    } @catch (NSException *exception) {
        reject(@"MQTTClientSide",@"send Message expection",nil);
    }
}

RCT_REMAP_METHOD(receiveMessage,receiveMessageWithId:(NSString*)sideId queueName:(NSString*)queueName resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            while (1) {
                @try {
                    MQTTClientSide* side = [JSObjManager getObjWithKey:sideId];
                    NSString* str1 = nil,* str2 = nil;
                    [side receiveMessage:&str1 message:&str2];
                    if (str1!=nil&&str2!=nil) {
                        [self sendEventWithName:queueName body:@{@"topic":str1,@"message":str2}];
//                    NSDictionary* dic = @{@"topic":str1,@"message":str2};
//                    resolve(@{@"message":dic});
                        [NSThread sleepForTimeInterval:3];
                    }
                } @catch (NSException *exception) {
                    reject(@"STOMPReceiver",@"receive Message expection",nil);
                }
            }
        });
    resolve(@"done");
}

RCT_REMAP_METHOD(subscribe,subscribeWithId:(NSString*)sideId topic:(NSString*)topicName qos:(int)qos resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        MQTTClientSide* side = [JSObjManager getObjWithKey:sideId];
        BOOL isSubscribe = [side subscribe:topicName quality:qos];
        NSNumber* num = [NSNumber numberWithBool:isSubscribe];
        resolve(@{@"isSubscribe":num});
    } @catch (NSException *exception) {
        reject(@"MQTTClientSide",@"subscribe expection",nil);
    }
}

RCT_REMAP_METHOD(unsubscribe,unsubscribeWithId:(NSString*)sideId topic:(NSString*)topicName resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        MQTTClientSide* side = [JSObjManager getObjWithKey:sideId];
        BOOL unSubscribe = [side unsubscribe:topicName];
        NSNumber* num = [NSNumber numberWithBool:unSubscribe];
        resolve(@{@"unSubscribe":num});
    } @catch (NSException *exception) {
        reject(@"MQTTClientSide",@"unSubscribe expection",nil);
    }
}

RCT_REMAP_METHOD(suspend,suspendWithId:(NSString*)sideId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        MQTTClientSide* side = [JSObjManager getObjWithKey:sideId];
        [side suspend];
        resolve(@"suspend");
    } @catch (NSException *exception) {
        reject(@"MQTTClientSide",@"suspend expection",nil);
    }
}

RCT_REMAP_METHOD(resume,resumeWithId:(NSString*)sideId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        MQTTClientSide* side = [JSObjManager getObjWithKey:sideId];
        BOOL isResume = [side resume];
        NSNumber* num = [NSNumber numberWithBool:isResume];
        resolve(@{@"isResume":num});
    } @catch (NSException *exception) {
        reject(@"MQTTClientSide",@"resume expection",nil);
    }
}

RCT_REMAP_METHOD(dispose,disposeWithId:(NSString*)sideId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        MQTTClientSide* side = [JSObjManager getObjWithKey:sideId];
        [JSObjManager removeObj:sideId];
        [side dispose];
        resolve(@"dispose");
    } @catch (NSException *exception) {
        reject(@"MQTTClientSide",@"dispose expection",nil);
    }
}
@end
